aclk/aclk_lws_wss_client.c

```diff
@@ -214,76 +245,86 @@ aclk_lws_wss_callback(struct lws *wsi, enum lws_callback_reasons reason,
 		inst->upstream_reconnect_request = 0;
 	}
 
+	// Don't log to info - volume is proportional to message flow on ACLK.
+	switch (reason) {
+		case LWS_CALLBACK_CLIENT_WRITEABLE:
+			if (inst->aborting)
+				return -1;
+			aclk_lws_mutex_lock(&inst->write_buf_mutex);
+			data = lws_wss_packet_buffer_pop(&inst->write_buffer_head);
+			if(likely(data)) {
+				lws_write(wsi, data->data + LWS_PRE, data->data_size, LWS_WRITE_BINARY);
+				lws_wss_packet_buffer_free(data);
+				if(inst->write_buffer_head)
+					lws_callback_on_writable(inst->lws_wsi);
+			}
+			aclk_lws_mutex_unlock(&inst->write_buf_mutex);
+			return retval;
+
+		case LWS_CALLBACK_CLIENT_RECEIVE:
+			aclk_lws_mutex_lock(&inst->read_buf_mutex);
+			if(!received_data_to_ringbuff(inst->read_ringbuffer, in, len))
+				retval = 1;
+			aclk_lws_mutex_unlock(&inst->read_buf_mutex);
+
+			if(likely(inst->callbacks.data_rcvd_callback))
+				// to future myself -> do not call this while read lock is active as it will eventually
+				// want to acquire same lock later in aclk_lws_wss_client_read() function
+				inst->callbacks.data_rcvd_callback();
+			else
+				inst->data_to_read = 1; //to inform logic above there is reason to call mosquitto_loop_read
```

Comment:
Haven’t looked at the mosquito layer, but for safety shouldn’t this be part of the critical section?